cdc/capture/capture.go

```diff
@@ -0,0 +1,332 @@
+		etcdClient: etcdClient,
+
+		newProcessorManager: processor.NewManager,
+		newOwner:            owner.NewOwner,
+	}
+}
+
+func (c *Capture) reset() error {
+	c.captureMu.Lock()
+	defer c.captureMu.Unlock()
+	conf := config.GetGlobalServerConfig()
+	c.info = &model.CaptureInfo{
+		ID:            uuid.New().String(),
+		AdvertiseAddr: conf.AdvertiseAddr,
+		Version:       version.ReleaseVersion,
+	}
+	c.processorManager = c.newProcessorManager()
+	if c.session != nil {
+		c.session.Close() //nolint:errcheck
+	}
+	sess, err := concurrency.NewSession(c.etcdClient.Client.Unwrap(),
+		concurrency.WithTTL(conf.CaptureSessionTTL))
+	if err != nil {
+		return errors.Annotate(cerror.WrapError(cerror.ErrNewCaptureFailed, err), "create capture session")
+	}
+	c.session = sess
+	c.election = concurrency.NewElection(sess, kv.CaptureOwnerKey)
+	atomic.StoreInt32(&c.stopCampaign, 0)
+	log.Info("init capture", zap.String("capture-id", c.info.ID), zap.String("capture-addr", c.info.AdvertiseAddr))
+	return nil
+}
+
+// Run runs the capture
+func (c *Capture) Run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+		err := c.reset()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err = c.run(ctx)
+		// if no error returned or capture suicide, reset the capture and run again
+		if err == nil || cerror.ErrCaptureSuicide.Equal(errors.Cause(err)) {
+			log.Info("capture recovered", zap.String("capture-id", c.info.ID))
+			continue
+		}
+		return errors.Trace(err)
+	}
+}
+
+func (c *Capture) run(stdCtx context.Context) error {
+	ctx := cdcContext.NewContext(stdCtx, &cdcContext.GlobalVars{
+		PDClient:    c.pdClient,
+		KVStorage:   c.kvStorage,
+		CaptureInfo: c.info,
+		EtcdClient:  c.etcdClient,
+	})
+	err := c.register(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer func() {
+		timeoutCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		if err := ctx.GlobalVars().EtcdClient.DeleteCaptureInfo(timeoutCtx, c.info.ID); err != nil {
+			log.Warn("failed to delete capture info when capture exited", zap.Error(err))
+		}
+		cancel()
+	}()
+	wg := new(sync.WaitGroup)
+	wg.Add(2)
+	var ownerErr, processorErr error
+	go func() {
+		defer log.Info("owner goroutine is exited")
+		defer wg.Done()
+		defer c.AsyncClose()
+		ownerErr = c.campaignOwner(ctx)
+	}()
+	go func() {
+		defer log.Info("processor goroutine is exited")
+		defer wg.Done()
+		defer c.AsyncClose()
+		conf := config.GetGlobalServerConfig()
+		processorFlushInterval := time.Duration(conf.ProcessorFlushInterval)
+		// no matter why the processor is exited, return to let capture restart or exit
+		processorErr = c.runEtcdWorker(ctx, c.processorManager, model.NewGlobalState(), processorFlushInterval)
+	}()
+	wg.Wait()
+	if ownerErr != nil {
+		return errors.Annotate(ownerErr, "owner exited with error")
+	}
+	if processorErr != nil {
+		return errors.Annotate(processorErr, "processor exited with error")
+	}
```

Comment:
There are two independent goroutines above, the `owner` routine and the `processorManager` routine, should the error in one of the goroutines terminate the running of the other one?
To see it in another way, could it happen that one of the goroutine exits and the other one is still running?